httpcore5/src/test/java/org/apache/hc/core5/http/impl/io/TestDefaultResponseOutOfOrderStrategy.java

```diff
@@ -0,0 +1,244 @@
+/*
+ * ====================================================================
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * ====================================================================
+ *
+ * This software consists of voluntary contributions made by many
+ * individuals on behalf of the Apache Software Foundation.  For more
+ * information on the Apache Software Foundation, please see
+ * <http://www.apache.org/>.
+ *
+ */
+
+package org.apache.hc.core5.http.impl.io;
+
+import org.apache.hc.core5.http.ClassicHttpRequest;
+import org.apache.hc.core5.http.message.BasicClassicHttpRequest;
+import org.apache.hc.core5.util.Timeout;
+import org.junit.Assert;
+import org.junit.Test;
+
+import javax.net.ssl.SSLSession;
+import javax.net.ssl.SSLSessionContext;
+import javax.security.cert.X509Certificate;
+import java.io.IOException;
+import java.io.InputStream;
+import java.security.Principal;
+import java.security.cert.Certificate;
+
+public class TestDefaultResponseOutOfOrderStrategy {
+
+    private static final ClassicHttpRequest REQUEST = new BasicClassicHttpRequest("POST", "/path");
+
+    @Test
+    public void testFirstChunkIsCheckedSsl() throws IOException {
+        final boolean earlyResponse = DefaultResponseOutOfOrderStrategy.INSTANCE.checkForEarlyResponse(
+                new StubConnection(true, true),
+                REQUEST,
+                // SSLSocket streams report zero bytes available
+                new StubInputStream(0),
+                0,
+                1);
+        Assert.assertTrue(earlyResponse);
+    }
+
+    @Test
+    public void testFirstChunkIsCheckedPlain() throws IOException {
+        final boolean earlyResponse = DefaultResponseOutOfOrderStrategy.INSTANCE.checkForEarlyResponse(
+                new StubConnection(true, false),
```

Comment:
@carterkozak Any reason for using those `Stub`s instead of Mockito mocks? `StubSSLSession` looks particularly ugly.

Reply:
I'm not sure why I assumed mockito wasn't available, sorry about that.